File: Algo/Leetcode/912Sorting.cpp
// 1. Use QuickSort(O(NlogN)). Remember to use random pivot rather than always the last one.
class Solution {
public:
    vector<int> sortArray(vector<int>& nums) {
        sortArray(nums, 0, nums.size()-1);
        return nums;
    }

    void sortArray(vector<int>& nums, int start, int end){
        if (start < end){
            int p = random_partition(nums, start, end);
            sortArray(nums, start, p-1);
            sortArray(nums, p+1, end);
        }
    }

    int random_partition(vector<int>& nums, int start, int end){
        int pivot = rand() % (end - start + 1) + start;
        swap(nums[pivot], nums[end]);
        int p = partition(nums, start, end);
        return p;
    }

    int partition(vector<int>& nums, int start, int end){
        int pivot = nums[end];
        int move = start - 1;
        for(int i=start; i<end; i++){
            if (nums[i] <= pivot)
                swap(nums[++move], nums[i]);
        }
        swap(nums[++move], nums[end]);
        return move;
    }
};

File: Algo/Leetcode/912Sorting.py
class Solution(object):
    def sortArray(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        self.sort(nums, 0, len(nums)-1)

        return nums

    def sort(self, nums, l, r):
        if l < r:
            pivot = self.partition(nums, l, r)
            self.sort(nums, l, pivot-1)
            self.sort(nums, pivot+1, r)

    def partition(self, nums, l, r):
        pivot = self.random_idx(l, r)
        self.swap(nums, pivot, r)

        pivot = l
        for i in range(l, r):
            if nums[i] < nums[r]:
                self.swap(nums, i, pivot)
                pivot += 1

        self.swap(nums, pivot, r)

        return pivot

    def random_idx(self, l, r):
        return random.randint(l, r)

    def swap(self, nums, i, j):
        tmp = nums[i]
        nums[i] = nums[j]
        nums[j] = tmp


# This do not pass the case of 2222.....2222, need two pointer